alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,121 @@
+
+use crate::tty::ChildEvent;
+
+/// WinAPI callback to run when child process exits.
+extern "system" fn child_exit_callback(ctx: PVOID, timed_out: BOOLEAN) {
+    if timed_out != 0 {
+        return;
+    }
+
+    let sender: Box<_> = unsafe { Box::from_raw(ctx as *mut Sender<ChildEvent>) };
+    let _ = sender.send(ChildEvent::Exited);
+}
+
+pub struct ChildExitWatcher {
+    wait_handle: AtomicPtr<c_void>,
+    event_rx: Receiver<ChildEvent>,
+}
+
+impl ChildExitWatcher {
+    pub fn new(child_handle: HANDLE) -> Result<ChildExitWatcher, Error> {
+        let (sender, receiver) = channel::<ChildEvent>();
+
+        let mut wait_handle: HANDLE = 0 as HANDLE;
+        let sender_ref = Box::new(sender);
+
+        let success = unsafe {
+            RegisterWaitForSingleObject(
+                &mut wait_handle,
+                child_handle,
+                Some(child_exit_callback),
+                Box::into_raw(sender_ref) as PVOID,
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if success == 0 {
+            Err(Error::last_os_error())
+        } else {
+            Ok(ChildExitWatcher { wait_handle: AtomicPtr::from(wait_handle), event_rx: receiver })
+        }
+    }
+
+    pub fn event_rx(&self) -> &Receiver<ChildEvent> {
+        &self.event_rx
+    }
+}
+
+impl Drop for ChildExitWatcher {
+    fn drop(&mut self) {
+        unsafe {
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::os::windows::io::AsRawHandle;
+    use std::process::Command;
+    use std::time::Duration;
+
+    use mio::{Events, Poll, PollOpt, Ready, Token};
+
+    use super::*;
+
+    #[test]
+    pub fn event_is_emitted_when_child_exits() {
+        const WAIT_TIMEOUT: Duration = Duration::from_millis(200);
+
+        let (sender, receiver) = channel::<()>();
+        let mut command = Command::new("cmd.exe");
+        let mut child = command.spawn().unwrap();
+        let child_exit_watcher = ChildExitWatcher::new(child.as_raw_handle()).unwrap();
+
+        let mut poll = Poll::new().unwrap();
+        let mut events = Events::with_capacity(8);
+        let child_events_token = Token::from(0usize);
+
+        poll.register(
+            child_exit_watcher.event_rx(),
+            child_events_token,
+            Ready::readable(),
+            PollOpt::oneshot(),
+        )
+        .unwrap();
+
+        child.kill().unwrap();
+
+        // Poll for the event or fail with timeout if nothing has been sent
+        poll.poll(&mut events, Some(WAIT_TIMEOUT)).unwrap();
+        for event in events.iter() {
+            if event.token() == child_events_token {
+                // Verify the right event was emitted.
+                assert_eq!(child_exit_watcher.event_rx().try_recv().unwrap(), ChildEvent::Exited);
+                return;
+            }
+        }
+    }
```

Comment:
This doesn't correctly assert success. If there is no token that matches `child_events_token`, it will never assert anything. What other events could this receive, do we even need this?

In any case, we'd probably want to do something more akin to iterator logic here, like `events.iter().filter().next()` or something, that would make the code cleaner. Then we can just unwrap on the `next` and we'll handle the case that there's nothing too.

Reply:
Not entirely sure it is that simple, events just tells us that `event_rx` is ready to be read. Have a look how I reworked it and let me know if the solution acceptable.

Reply:
> events just tells us that event_rx is ready to be read

The `poll.poll` function will block until at least one readiness event is received. So `event_rx` should always be ready to be read.